File: consistency-store/src/main/java/de/tudarmstadt/consistency/store/Ref.java
package de.tudarmstadt.consistency.store;

/**
 * A handle corresponds to a single database object. It can be used to read the specified object or to write
 * a new value as that object.
 *
 * @author Mirko Köhler
 */
public interface Ref<T, R extends Ref<T, R>> {

	/**
	 * Handles an event for the database object represented by this handle.
	 * Examples for events are reading from the database or writing to the database.
	 * The concrete allowed events are defined by the database.
	 *
	 * @param e The operation that should be handled.
	 * @param param The argument to the operation.
	 * @return A value that is produced by the event, or null if the event does not
	 * produce any value.
	 *
	 * @throws Exception if the event can not be handled, or communication with
	 * the database goes wrong.
	 */
	<X, Y> Y handle(Operation<T, R, X, Y> e, X param) throws Exception;

}

File: consistency-store/src/main/scala/de/tudarmstadt/consistency/store/scala/Ref.scala
package de.tudarmstadt.consistency.store.scala

import scala.language.higherKinds

/**
	* A handle that corresponds to a key in the database. The handle is used to
	* perform operations on the referenced value.
	*
	* @tparam T the type of value that is referenced by this reference
	* @tparam Op the type of operations that can be performed by this reference
	*
	* @author Mirko Köhler
	*/
trait Ref[+T, Op[_,_] <: Operation[_,_]] {

	def handle[Parameter, Arity](op : Op[Parameter, Arity], parameter : Parameter) : Arity
}
